src/coprocessor/dag/rpn_expr/impl_cast.rs

```diff
@@ -164,54 +164,44 @@ pub fn cast_int_as_decimal(
 }
 
 macro_rules! cast_as_integer {
-    ($ty:ty, $as_int_fn:ident, $as_int_conv:ident) => {
-        cast_as_integer!($ty, $as_int_fn, $as_int_conv, val);
+    ($ty:ty, $as_int_fn:ident) => {
+        cast_as_integer!($ty, $as_int_fn, val);
     };
-    ($ty:ty, $as_int_fn:ident, $as_int_conv:ident, $expr:expr) => {
+    ($ty:ty, $as_int_fn:ident, $expr:expr) => {
         #[rpn_fn(capture = [ctx])]
         #[inline]
         pub fn $as_int_fn(ctx: &mut EvalContext, val: &Option<$ty>) -> Result<Option<i64>> {
             match val {
                 None => Ok(None),
                 Some(val) => {
-                    let val = $as_int_conv(ctx, $expr, FieldTypeTp::LongLong)?;
+                    let val = $expr.to_int(ctx, FieldTypeTp::LongLong)?;
                     Ok(Some(val))
                 }
             }
         }
     };
 }
 
-cast_as_integer!(Int, cast_int_as_int, convert_int_to_int, *val);
-cast_as_integer!(Int, cast_uint_as_int, convert_uint_to_int, *val as u64);
-cast_as_integer!(
-    Real,
-    cast_float_as_int,
-    convert_float_to_int,
-    val.into_inner()
-);
-cast_as_integer!(Bytes, cast_bytes_as_int, convert_bytes_to_int);
-cast_as_integer!(Decimal, cast_decimal_as_int, convert_decimal_to_int);
-cast_as_integer!(DateTime, cast_datetime_as_int, convert_datetime_to_int);
-cast_as_integer!(
-    Duration,
-    cast_duration_as_int,
-    convert_duration_to_int,
-    *val
-);
-cast_as_integer!(Json, cast_json_as_int, convert_json_to_int);
+cast_as_integer!(Int, cast_int_as_int, *val);
+cast_as_integer!(Int, cast_uint_as_int, *val as u64);
+cast_as_integer!(Real, cast_float_as_int, val.into_inner());
+cast_as_integer!(Bytes, cast_bytes_as_int);
```

Comment:
Looks like this will be the benefit of introducing our own Cast trait from any type to any type (not only ToInt), because then we don't need:

cast_any_as_int
cast_any_as_real
cast_any_as_bytes
...

we will only need:

cast_any_as_any<SRC, DEST>